x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -278,9 +291,111 @@ private void maybeUpdateMappings(Client localClient, Client remoteClient, Index
+    private static class RestoreSession extends FileRestoreContext implements Closeable {
+
+        private static final int BUFFER_SIZE = 1 << 16;
+
+        private final Client remoteClient;
+        private final String sessionUUID;
+        private final DiscoveryNode node;
+        private final Store store;
+        private final Store.MetadataSnapshot sourceMetaData;
+
+        RestoreSession(String repositoryName, Client remoteClient, String sessionUUID, DiscoveryNode node, IndexShard indexShard,
+                       RecoveryState recoveryState, Store.MetadataSnapshot sourceMetaData) {
+            super(repositoryName, indexShard, SNAPSHOT_ID, recoveryState, BUFFER_SIZE);
+            this.remoteClient = remoteClient;
+            this.sessionUUID = sessionUUID;
+            this.node = node;
+            this.store = indexShard.store();
+            this.store.incRef();
+            this.sourceMetaData = sourceMetaData;
+        }
+
+        static RestoreSession openSession(String repositoryName, Client remoteClient, ShardId leaderShardId, IndexShard indexShard,
+                                          RecoveryState recoveryState) {
+            String sessionUUID = UUIDs.randomBase64UUID();
+            PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse response = remoteClient.execute(PutCcrRestoreSessionAction.INSTANCE,
+                new PutCcrRestoreSessionRequest(sessionUUID, leaderShardId)).actionGet();
+            return new RestoreSession(repositoryName, remoteClient, sessionUUID, response.getNode(), indexShard, recoveryState,
+                response.getStoreFileMetaData());
+        }
+
+        void restoreFiles() throws IOException {
+            ArrayList<BlobStoreIndexShardSnapshot.FileInfo> fileInfos = new ArrayList<>();
+            for (StoreFileMetaData fileMetaData : sourceMetaData) {
+                ByteSizeValue fileSize = new ByteSizeValue(fileMetaData.length());
+                fileInfos.add(new BlobStoreIndexShardSnapshot.FileInfo(fileMetaData.name(), fileMetaData, fileSize));
+            }
+            SnapshotFiles snapshotFiles = new SnapshotFiles(LATEST, fileInfos);
+            restore(snapshotFiles);
+        }
+
+        @Override
+        protected InputStream fileInputStream(BlobStoreIndexShardSnapshot.FileInfo fileInfo) {
+            return new RestoreFileInputStream(remoteClient, sessionUUID, node, fileInfo.metadata());
+        }
+
+        @Override
+        public void close() {
+            this.store.decRef();
+            ClearCcrRestoreSessionRequest clearRequest = new ClearCcrRestoreSessionRequest(sessionUUID, node);
+            ClearCcrRestoreSessionAction.ClearCcrRestoreSessionResponse response =
+                remoteClient.execute(ClearCcrRestoreSessionAction.INSTANCE, clearRequest).actionGet();
+        }
+    }
+
+    private static class RestoreFileInputStream extends InputStream {
```

Comment:
would it make sense to use `SlicedInputStream` instead here? For example, something like

```
private static class RestoreFileInputStream extends SlicedInputStream {

        private final Client remoteClient;
        private final String sessionUUID;
        private final DiscoveryNode node;
        private final StoreFileMetaData fileToRecover;

        private RestoreFileInputStream(Client remoteClient, String sessionUUID, DiscoveryNode node, StoreFileMetaData fileToRecover) {
            super((fileToRecover.length() / BUFFER_SIZE) + 1);
            this.remoteClient = remoteClient;
            this.sessionUUID = sessionUUID;
            this.node = node;
            this.fileToRecover = fileToRecover;
        }

        @Override
        protected InputStream openSlice(long slice) throws IOException {
            long pos = BUFFER_SIZE * slice;
            int bytesRequested = Math.toIntExact(Math.min(BUFFER_SIZE, fileToRecover.length() - pos));
            String fileName = fileToRecover.name();
            GetCcrRestoreFileChunkRequest request = new GetCcrRestoreFileChunkRequest(node, sessionUUID, fileName, pos, bytesRequested);
            BytesReference fileChunk = remoteClient.execute(GetCcrRestoreFileChunkAction.INSTANCE, request).actionGet().getChunk();
            return fileChunk.streamInput();
        }
    }
```

Reply:
I'm not sure? The `SlicedInputStream` would depend on the fact that the other side always sends the total number of bytes requested as we move on to the next slice.  This may not be a correct assumption depending on how me might implement rate limiting on the leader side? We may also want to have the flexibility to not observe the requested buffer size on the leader side?

I did adopt your change of using the byte reference input stream as that is simpler.

Reply:
rate limiting for peer recovery currently inserts `Thread.sleeps`, but always sticks with the same chunk size. We can revisit this later.